### PR TITLE
feat: Part 2 - Add validation checks for TuningSpec, DataSource, DataDestination

### PR DESCRIPTION
**Pull request by ishaansehgal99**
Open to feedback. Here is validation checks for the new CRD.

Code Coverage: 87.7% - workspace_validation.go

[coverage.txt](https://github.com/Azure/kaito/files/14661460/coverage.txt)

### CODE REVIEW

File: api/v1alpha1/workspace_validation.go [hunk 1]
```diff
@@ -35,20 +35,152 @@ func (w *Workspace) Validate(ctx context.Context) (errs *apis.FieldError) {
...
+			w.Resource.validateUpdate(&old.Resource).ViaField("resource"),
 		)
 	}
 	return errs
 }
 
+func (w *Workspace) validateCreate() (errs *apis.FieldError) {
+	inferenceSpecified := w.Inference.Preset != nil || w.Inference.Template != nil
+	tuningSpecified := w.Tuning.Input != nil
+	if inferenceSpecified != tuningSpecified {
```
Comment (Fei-Guo) [inline api/v1alpha1/workspace_validation.go:63]: Is this wrong? We need to add UT?

File: api/v1alpha1/workspace_validation.go [hunk 2]
```diff
@@ -35,20 +35,152 @@ func (w *Workspace) Validate(ctx context.Context) (errs *apis.FieldError) {
...
+		errs = errs.Also(apis.ErrMissingField("Input"))
+	}
+	if r.Output == nil {
+		errs = errs.Also(apis.ErrMissingField("Output"))
+	}
+	// Currently require a preset to specified, in future we can consider defining a template
+	if r.Preset == nil {
+		errs = errs.Also(apis.ErrMissingField("Preset"))
+	}
+	// TODO: We have to register training plugins and check if it preset exists in plugins here
```
Comment (Fei-Guo) [inline api/v1alpha1/workspace_validation.go:97]: isValidPreset, we have this method, right?

File: api/v1alpha1/workspace_validation.go [hunk 3]
```diff
@@ -35,20 +35,152 @@ func (w *Workspace) Validate(ctx context.Context) (errs *apis.FieldError) {
...
+	}
+	if !reflect.DeepEqual(old.Output, r.Output) {
+		errs = errs.Also(apis.ErrGeneric("Output field cannot be changed", "Output"))
+	}
+	if !reflect.DeepEqual(old.Preset, r.Preset) {
+		errs = errs.Also(apis.ErrGeneric("Preset cannot be changed", "Preset"))
+	}
+	// We will have to consider supporting tuning method and config fields changing
+	methodLowerCase := strings.ToLower(string(r.Method))
+	if methodLowerCase != string(TuningMethodLora) && methodLowerCase != string(TuningMethodQLora) {
```
Comment (Fei-Guo) [inline api/v1alpha1/workspace_validation.go:117]: We should not even allow changing the method.

File: api/v1alpha1/workspace_validation.go [hunk 4]
```diff
@@ -35,20 +35,152 @@ func (w *Workspace) Validate(ctx context.Context) (errs *apis.FieldError) {
...
+	// Ensure exactly one of URLs, HostPath, or Image is specified
+	if sourcesSpecified != 1 {
+		errs = errs.Also(apis.ErrGeneric("Exactly one of URLs, HostPath, or Image must be specified", "URLs", "HostPath", "Image"))
+	}
+
+	return errs
+}
+
+func (r *DataSource) validateUpdate(old *DataSource) (errs *apis.FieldError) {
+	if !reflect.DeepEqual(old.URLs, r.URLs) {
```
Comment (Fei-Guo) [inline api/v1alpha1/workspace_validation.go:144]: You'd better sorting arrays first, then applying DeepEqual.

File: api/v1alpha1/workspace_validation.go [hunk 5]
```diff
@@ -35,20 +35,152 @@ func (w *Workspace) Validate(ctx context.Context) (errs *apis.FieldError) {
...
+		errs = errs.Also(apis.ErrInvalidValue("URLs field cannot be changed once set", "URLs"))
+	}
+	if old.HostPath != r.HostPath {
+		errs = errs.Also(apis.ErrInvalidValue("HostPath field cannot be changed once set", "HostPath"))
+	}
+	if old.Image != r.Image {
+		errs = errs.Also(apis.ErrInvalidValue("Image field cannot be changed once set", "Image"))
+	}
+	// TODO: Ensure ImageSecrets can be changed
```
Comment (Fei-Guo) [inline api/v1alpha1/workspace_validation.go:153]: can -> cannot

File: api/v1alpha1/workspace_validation.go [hunk 6]
```diff
@@ -35,20 +35,152 @@ func (w *Workspace) Validate(ctx context.Context) (errs *apis.FieldError) {
...
+}
+
+func (r *DataDestination) validateUpdate(old *DataDestination) (errs *apis.FieldError) {
+	if old.HostPath != r.HostPath {
+		errs = errs.Also(apis.ErrInvalidValue("HostPath field cannot be changed once set", "HostPath"))
+	}
+	if old.Image != r.Image {
+		errs = errs.Also(apis.ErrInvalidValue("Image field cannot be changed once set", "Image"))
+	}
+	// TODO: Ensure ImageSecrets can be changed
```
Comment (Fei-Guo) [inline api/v1alpha1/workspace_validation.go:180]: can -> cannot

**Review by Fei-Guo — COMMENTED** [review]
Please add sufficient UTs.

File: api/v1alpha1/workspace_validation.go (hunk 1 above)
Comment (ishaansehgal99) [reply api/v1alpha1/workspace_validation.go]: Yes this was mistake, thanks UT added

File: api/v1alpha1/workspace_validation.go (hunk 3 above)
Comment (ishaansehgal99) [reply api/v1alpha1/workspace_validation.go]: done

File: api/v1alpha1/workspace_validation.go (hunk 5 above)
Comment (ishaansehgal99) [reply api/v1alpha1/workspace_validation.go]: done

File: api/v1alpha1/workspace_validation.go (hunk 4 above)
Comment (ishaansehgal99) [reply api/v1alpha1/workspace_validation.go]: done

File: api/v1alpha1/workspace_validation.go (hunk 6 above)
Comment (ishaansehgal99) [reply api/v1alpha1/workspace_validation.go]: done

File: api/v1alpha1/workspace_validation.go (hunk 2 above)
Comment (ishaansehgal99) [reply api/v1alpha1/workspace_validation.go]: Yep. Thing is if someone uses an inference preset for performing training? Something to think about

Comment (Fei-Guo) [reply api/v1alpha1/workspace_validation.go]: We only support fine tuning the model that kaito supports for inference if user specifies the PresetSpec. When we start to support pod template in Tuning spec, the check can be by passed.

**Review by Fei-Guo — APPROVED** [review]

**Pull request merged**